File: app/src/features/files/filePresentation.ts
// Pure presentation logic for the file browser.
//
// Names are byte-faithful throughout: nothing here case-folds, normalizes
// NFC to NFD, or trims. Two entries that differ only by Unicode normalization
// are different files on the server, and collapsing them in the UI would send
// an operation to the wrong one.

import type { FileEntry } from '../../lib/files';
import type { IconName } from '../../components/ui/icon';

/** A stable list key. */
export function entryKey(dir: string, entry: FileEntry): string {
  // The full path, exactly as stored. Using the bare name would collide
  // between directories during a navigation, and normalizing would merge two
  // genuinely distinct entries.
  return `${dir}\u0000${entry.name}`;
}

export function entryIcon(entry: FileEntry): IconName {
  if (entry.isSymlink) return 'link-off';
  if (entry.isDir) return 'folder';
  return 'file';
}

export function formatSize(n: number): string {
  if (!Number.isFinite(n) || n < 0) return '';
  if (n < 1024) return `${n} B`;
  const units = ['KiB', 'MiB', 'GiB', 'TiB'];
  let v = n / 1024;
  let i = 0;
  while (v >= 1024 && i < units.length - 1) {
    v /= 1024;
    i += 1;
  }
  return `${v.toFixed(v >= 100 ? 0 : 1)} ${units[i]}`;
}

export function formatMtime(sec: number): string {
  if (sec <= 0) return '';
  return new Date(sec * 1000).toLocaleString();
}

export function entryDescription(entry: FileEntry): string {
  const parts: string[] = [];
  if (!entry.isDir) {
    const size = formatSize(entry.size);
    if (size !== '') parts.push(size);
  }
  const when = formatMtime(entry.mtime);
  if (when !== '') parts.push(when);
  return parts.join('  ');
}

/** A screen-reader label that states what the row is, not just its name. */
export function entryAccessibilityLabel(entry: FileEntry): string {
  const kind = entry.isSymlink ? 'link' : entry.isDir ? 'folder' : 'file';
  const detail = entryDescription(entry);
  return detail === ''
    ? `${kind} ${entry.name}`
    : `${kind} ${entry.name}, ${detail}`;
}

/** Why a proposed name cannot be used, or null when it is acceptable. */
export function validateName(name: string): string | null {
  if (name === '') return 'Enter a name.';
  if (name === '.' || name === '..') return 'That name is reserved.';
  if (name.includes('/') || name.includes('\\')) {
    return 'A name cannot contain a path separator.';
  }
  // A NUL terminates the string in most server-side APIs, so a name carrying
  // one would address a different file than the one displayed.
  if (name.includes('\u0000'))
    return 'That name contains an invalid character.';
  for (const c of name) {
    if (c.codePointAt(0)! < 0x20) {
      return 'That name contains a control character.';
    }
  }
  if (name.length > 255) return 'That name is too long.';
  return null;
}

/** True when a rename would collide with an entry already in the directory. */
export function nameExists(
  entries: readonly FileEntry[],
  name: string,
): boolean {
  // Exact comparison. A server that distinguishes two normalization forms has
  // two files, and treating them as one would overwrite the wrong one.
  return entries.some(e => e.name === name);
}

export type TransferDirection = 'upload' | 'download';
export type TransferState =
  | 'queued'
  | 'running'
  | 'done'
  | 'failed'
  | 'cancelled'
  | 'conflict';

export interface Transfer {
  id: string;
  direction: TransferDirection;
  /** The name shown to the user. Not a local path. */
  displayName: string;
  remotePath: string;
  transferred: number;
  total: number;
  state: TransferState;
  /** True when this backend can resume an interrupted transfer. */
  resumable: boolean;
  error?: string;
}

/** Progress as a fraction, or null when the total is unknown. */
export function transferProgress(t: Transfer): number | null {
  if (t.total <= 0) return null;
  return Math.min(1, Math.max(0, t.transferred / t.total));
}

export function transferLabel(t: Transfer): string {
  const verb = t.direction === 'upload' ? 'Uploading' : 'Downloading';
  switch (t.state) {
    case 'queued':
      return `Waiting to ${t.direction}`;
    case 'running':
      return `${verb} ${t.displayName}`;
    case 'done':
      return `${t.direction === 'upload' ? 'Uploaded' : 'Downloaded'} ${
        t.displayName
      }`;
    case 'failed':
      return `${t.displayName} failed`;
    case 'cancelled':
      return `${t.displayName} cancelled`;
    case 'conflict':
      return `${t.displayName} already exists`;
    default:
      return t.displayName;
  }
}

/** How a name conflict is resolved. Overwrite is never the default. */
export type ConflictChoice = 'cancel' | 'rename' | 'overwrite';

/**
 * A non-colliding name, by inserting a counter before the extension.
 *
 * `report.txt` becomes `report (1).txt`, matching what a desktop file manager
 * does, so the result is recognizable rather than a timestamped surprise.
 */
/**
 * Splits a name into the stem and extension a counter is inserted between.
 *
 * A leading dot is a hidden file, not an extension, so `.bashrc` keeps its
 * whole name as the stem and becomes `.bashrc (1)`.
 */
export function splitName(name: string): { stem: string; ext: string } {
  const dot = name.lastIndexOf('.');
  if (dot <= 0) return { stem: name, ext: '' };
  return { stem: name.slice(0, dot), ext: name.slice(dot) };
}

/** The nth candidate name, matching what a desktop file manager produces. */
export function numberedName(name: string, n: number): string {
  const { stem, ext } = splitName(name);
  return `${stem} (${n})${ext}`;
}

export function uniqueName(
  entries: readonly FileEntry[],
  name: string,
): string {
  if (!nameExists(entries, name)) return name;
  for (let i = 1; i < 1000; i += 1) {
    const candidate = numberedName(name, i);
    if (!nameExists(entries, candidate)) return candidate;
  }
  const { stem, ext } = splitName(name);
  return `${stem} (${Date.now()})${ext}`;
}

File: app/android/app/src/main/java/com/remotly/app/files/FilePresentation.kt
package com.remotly.app.files

import java.text.DateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

/**
 * Presentation for the file browser.
 *
 * Names are byte-faithful throughout: nothing here case-folds, normalizes
 * NFC to NFD, or trims. Two entries differing only by Unicode normalization
 * are different files on the server, and collapsing them in the UI would
 * send an operation to the wrong one.
 */

fun formatSize(n: Long): String {
    if (n < 0) return ""
    if (n < 1024) return "$n B"
    val units = listOf("KiB", "MiB", "GiB", "TiB")
    var v = n.toDouble() / 1024
    var i = 0
    while (v >= 1024 && i < units.size - 1) {
        v /= 1024
        i += 1
    }
    return if (v >= 100) {
        String.format(Locale.getDefault(), "%.0f %s", v, units[i])
    } else {
        String.format(Locale.getDefault(), "%.1f %s", v, units[i])
    }
}

/**
 * A compact timestamp for a list row.
 *
 * A full locale date and time is roughly twenty characters and pushes the
 * size off the end of a narrow row. A file manager shows the time for today,
 * the day and month within the year, and the year beyond that, which is
 * enough to tell two versions of a file apart at a glance.
 *
 * [now] is injectable so the boundaries are testable.
 */
fun formatMtime(seconds: Long, now: Date = Date()): String {
    if (seconds <= 0) return ""
    val at = Date(seconds * 1000)
    val a = Calendar.getInstance().apply { time = at }
    val b = Calendar.getInstance().apply { time = now }
    val sameYear = a.get(Calendar.YEAR) == b.get(Calendar.YEAR)
    val sameDay = sameYear && a.get(Calendar.DAY_OF_YEAR) == b.get(Calendar.DAY_OF_YEAR)
    return when {
        sameDay -> DateFormat.getTimeInstance(DateFormat.SHORT).format(at)
        sameYear -> DateFormat.getDateInstance(DateFormat.MEDIUM).format(at)
        else -> DateFormat.getDateInstance(DateFormat.MEDIUM).format(at)
    }
}

fun entryDescription(entry: FileEntry, now: Date = Date()): String {
    val parts = mutableListOf<String>()
    if (!entry.isDir) {
        val size = formatSize(entry.size)
        if (size.isNotEmpty()) parts.add(size)
    }
    val when_ = formatMtime(entry.mtime, now)
    if (when_.isNotEmpty()) parts.add(when_)
    return parts.joinToString("  ")
}

/** A screen-reader label that states what the row is, not just its name. */
fun entryAccessibilityLabel(entry: FileEntry, now: Date = Date()): String {
    val kind = when {
        entry.isSymlink -> "link"
        entry.isDir -> "folder"
        else -> "file"
    }
    val detail = entryDescription(entry, now)
    return if (detail.isEmpty()) "$kind ${entry.name}" else "$kind ${entry.name}, $detail"
}

/** Why a proposed name cannot be used, or null when it is acceptable. */
fun validateName(name: String): String? {
    val trimmed = name.trim()
    if (trimmed.isEmpty()) return "Enter a name."
    if (trimmed == "." || trimmed == "..") return "That name is reserved."
    if (trimmed.contains('/') || trimmed.contains('\\')) return "Use a name without a slash."
    if (trimmed.contains('\u0000')) return "That name contains an invalid character."
    return null
}

/** True when a rename would collide with an entry already in the directory. */
fun nameExists(entries: List<FileEntry>, name: String): Boolean =
    entries.any { it.name == name }

/** Splits a name into the stem and extension a counter is inserted between. */
fun splitName(name: String): Pair<String, String> {
    val dot = name.lastIndexOf('.')
    if (dot <= 0 || dot == name.length - 1) return name to ""
    return name.substring(0, dot) to name.substring(dot)
}

/** The nth candidate name, matching what a desktop file manager produces. */
fun numberedName(name: String, n: Int): String {
    val (stem, ext) = splitName(name)
    return "$stem ($n)$ext"
}

/** The first name free in [entries], starting from the one asked for. */
fun uniqueName(entries: List<FileEntry>, name: String): String {
    if (!nameExists(entries, name)) return name
    for (i in 1..999) {
        val candidate = numberedName(name, i)
        if (!nameExists(entries, candidate)) return candidate
    }
    return numberedName(name, System.currentTimeMillis().toInt())
}
